### ISSUE TITLE
More unified and structured HTML reporting with export to PDF

### ISSUE THREAD

**Issue by tiborauer** [automaticanalysis/automaticanalysis#244]
**Is your feature request related to a problem? Please describe.**
The current reporting does not give detailed information and needs more structure/level to reduce loading time and exporting facility. 

**Describe the solution you'd like**
- [ ] Omit steps with no report
- [ ] More text to certain steps
- [ ] An additional level for single-subject reports (e.g. grouping QA, registration, modelling)
- [ ] An intuitive outlier marking ('red flag') should also be implemented.
- [ ] The report should be exported in PDF(s).

**Comment by vijayiyer05** [comment]
Following up our discussion, one option might be the MATLAB Report Generator, which provides an abstract model for [content generation](https://www.mathworks.com/help/rptgen/content-generation.html) and allows for PDF, HTML, and other outputs

**Comment by AljenU** [comment]
On "The report should be exported in PDF"

Questions:
The goal is to make it easy to pass around a report? 
And one reason this is currently not easy is because the images in the report are links, that can be all over the directory structure?
Is it also acceptable to pass around a zipped html report (if that can be done such that what files it needs are all included), is that also an acceptable solution?

One the one hand, a zipped html report does not require layout changes, and will preserve links and image resolution.
On the other hand, a PDF might be more accessible.

It might be possible to go from html directly to pdf using the report generator (https://nl.mathworks.com/help/rptgen/ug/mlreportgen.dom.htmlfile-class.html) without first having to adapt the html content.

**Comment by tiborauer** [comment]
I agree, the zipped HTML (archive with all images) could be the "lowest hanging fruit". 

The HTML could be "printed" into PDF, but I am not sure how it could be automated.

The report generator seems to be the sleakest solution, but I have no experience with it, yet.

**Comment by AljenU** [comment]
Options w.r.t. exportable reports:

0. Put all the reports in a dedicated subfolder of the studyroot. With only the main report file as single file at the top level, and all other htm files in separate folders. Somewhat optional for create-pdf, but will provide a better view on what files and folders are part of the report, and which are not. Folder structure proposal example, assuming also images are copied:
- studyroot/REPORT_FOLDER_NAME
  - report.htm
  - report_firstlevel
    - report_firstlevel.htm
  - report_moco
    - report_moco.htm
    - report_moco_files
      - img_0001_QUALIFIEDNAME.jpeg
      - img_0002_QUALIFIEDNAME.jpeg
  - report_subjects
    - report_subjects.htm
    - report_S01.htm
    - report_S02.htm
    - report_S01_files
      - img_0001_QUALIFIEDNAME.jpg


1. Work on being able to provide a zipped-html report
1.a. Have all reports together in one subfolder (i.e. 0.). So in the end, zipping that one folder will contain all that is needed to show the html pages.
1.b. Have all the images / videos next to the report file that uses them. Assuming that all report generation uses aas_report_addimage to add images / videos to a report, adding a copy-then-add in this function should do the trick.
1.c. Make all local links relative. Both links to other report htm files, and links to images and videos.
1.d. Zip the folder (only when requested? have a setting?)

2. Work on using the report generator to create a pdf. Assumes we want 1 pdf with all the info.
  a. Get together a list of all the htm files that need to be part of the pdf (item 0. would help here). Determine in what order they should appear in the pdf. OR Generate one pdf file per htm file
  b. Create one mlreportgen.dom.HTMLFile object per file
    1. Only in 2020a and later: Use mlreportgen.utils.html2dom.prepHTMLFile to prepare the html files for adding as HTMLFile object
    2. If from 2015a and later: Fix the html code creation within aa so it creates html code that is compatible for use in a HTMLFile object. To do this, have a 2020a or later Matlab, and do a diff between the currently generated htm file, and when that htm file has been passed through mlreportgen.utils.html2dom.prepHTMLFile and saved as htm again. The diff will show what kind of changes are needed, and is needed because the error messages from trying to create a HTMLFile object from an non-compatible htm file are not informative enough. Combining information from the diff and the error messages can help to reduce the number of changes, since not all that is changed in the parsed htm file necessarily needs to be fixed.
  c. Add together all the HTMLFile objects into one big DOM object (and hope that it is not too big to fit in memory)
  d. Export the DOM object to pdf (Note that there are some Matlab documentation examples on combining dom objects, and on exporting dom objects)

**Comment by tiborauer** [comment]
It sounds very sensible. However, please, keep in mind that report.htm and report_firstlevel.htm also have images. Also, other use cases (e.g. M/EEG) may have other htms (also wth images).

Of course, the proposed solution seems easily extensible.
